File: util/lang/Formatter.java
package common.util.lang;

import common.CommonStatic;
import common.io.assets.Admin.StaticPermitted;
import common.io.json.JsonClass;
import common.io.json.JsonEncoder;
import common.io.json.JsonField;
import common.pack.Context.ErrType;
import common.pack.Identifier;
import common.util.pack.Background;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class Formatter {

	@JsonClass
	public static class Context {

		public final String left = "(";
		public final String right = ")";
		public final String sqleft = "[";
		public final String sqright = "]";
		public final String crleft = "{";
		public final String crright = "}";

		@JsonField
		public boolean isEnemy;
		@JsonField
		public boolean useSecond;

		public DecimalFormat df = new DecimalFormat("#.##");

		public Context() {
		}

		public Context(boolean ene, boolean sec) {
			isEnemy = ene;
			useSecond = sec;
		}

		public String abs(int v) {
			return "" + Math.abs(v);
		}

		public String bg(Identifier<Background> id) {
			return Identifier.get(id) + "";
		}

		public String dispTime(int time) {
			if (useSecond)
				return toSecond(time) + "s";
			return time + "f";
		}

		public String entity(Identifier<?> id) {
			return Identifier.get(id) + "";
		}

		public String toSecond(int time) {
			return df.format(time / 30.0);
		}

	}

	private class BoolElem extends Comp {

		public BoolElem(int start, int end) {
			super(start, end);
		}

		private boolean eval() throws Exception {
			for (int i = 0; i < MATCH.length; i++)
				for (int j = p0; j < p1 - MATCH[i].length() + 1; j++)
					if (test(i, j)) {
						int fi0 = new IntExp(p0, j).eval();
						int fi1 = new IntExp(j + MATCH[i].length(), p1).eval();
						if (i == 0)
							return fi0 >= fi1;
						if (i == 1)
							return fi0 <= fi1;
						if (i == 2)
							return fi0 == fi1;
						if (i == 3)
							return fi0 != fi1;
						if (i == 4)
							return fi0 > fi1;
						return fi0 < fi1;
					}
			return (Boolean) new RefObj(p0, p1).eval();
		}

		private boolean test(int i, int j) {
			for (int k = 0; k < MATCH[i].length(); k++)
				if (str.charAt(j + k) != MATCH[i].charAt(k))
					return false;
			return true;
		}

	}

	private class BoolExp extends Comp {

		private int ind;

		public BoolExp(int start, int end) {
			super(start, end);
			ind = p0;
		}

		private boolean eval() throws Exception {
			Stack<Boolean> stack = new Stack<>();
			stack.push(nextElem());
			while (ind < p1) {
				char ch = str.charAt(ind++);
				if (ch == '&')
					stack.push(stack.pop() & nextElem());
				else if (ch == '|')
					stack.push(nextElem());
				else
					throw new Exception("unknown operator " + ch + " at " + (ind - 1));
			}
			boolean b = false;
			for (boolean bi : stack)
				b |= bi;
			return b;
		}

		private boolean nextElem() throws Exception {
			char ch = str.charAt(ind);
			boolean neg = ch == '!';
			if (neg)
				ch = str.charAt(++ind);
			if (ch == '!')
				throw new Exception("double ! at " + ind);
			if (ch == '(') {
				int depth = 1;
				int pre = ++ind;
				while (depth > 0) {
					char chr = str.charAt(ind++);
					if (chr == '(')
						depth++;
					if (chr == ')')
						depth--;
				}
				return neg ^ new BoolExp(pre, ind - 1).eval();
			}
			int pre = ind;
			while (ch != '&' && ch != '|' && ind < p1)
				ch = str.charAt(++ind);
			return neg ^ new BoolElem(pre, ind).eval();
		}
	}

	private class Code implements IElem {

		private final BoolExp cond;
		private final Root data;

		private Code(BoolExp c, Root d) {
			cond = c;
			data = d;
		}

		@Override
		public void build(StringBuilder sb) throws Exception {
			if (cond.eval())
				data.build(sb);
		}

	}

	private class CodeBlock extends Cont {

		private CodeBlock(int start, int end) throws Exception {
			super(start, end);
			int i = p0;
			while (i < p1) {
				char ch = str.charAt(i++);
				if (ch == '(') {
					int depth = 1;
					int pre = i;
					while (depth > 0) {
						if (i >= p1)
							throw new Exception("unfinished at " + i);
						char chr = str.charAt(i++);
						if (chr == '(')
							depth++;
						if (chr == ')')
							depth--;
					}
					BoolExp cond = new BoolExp(pre, i - 1);
					while (str.charAt(i++) != '{')
						if (i >= p1)
							throw new Exception("unfinished at " + i);
					depth = 1;
					pre = i;
					while (depth > 0) {
						if (i >= p1)
							throw new Exception("unfinished at " + i);
						char chr = str.charAt(i++);
						if (chr == '{')
							depth++;
						if (chr == '}')
							depth--;
					}
					Root data = new Root(pre, i - 1);
					list.add(new Code(cond, data));
				}
			}
		}

	}

	private abstract static class Comp {

		public final int p0, p1;

		public Comp(int start, int end) {
			p0 = start;
			p1 = end;
		}
	}

	private abstract static class Cont extends Elem {

		public final List<IElem> list = new ArrayList<>();

		public Cont(int start, int end) {
			super(start, end);
		}

		@Override
		public void build(StringBuilder sb) throws Exception {
			for (IElem e : list)
				e.build(sb);
		}

	}

	private abstract static class Elem extends Comp implements IElem {

		public Elem(int start, int end) {
			super(start, end);
		}

	}

	private interface IElem {

		void build(StringBuilder sb) throws Exception;

	}

	private class IntExp extends Comp {

		private int ind;

		public IntExp(int start, int end) {
			super(start, end);
			ind = p0;
		}

		private int eval() throws Exception {
			Stack<Integer> stack = new Stack<>();
			stack.push(nextElem());
			char prevOp = ' ';
			while (ind < p1) {
				char ch = str.charAt(ind++);
				if (ch == '*')
					stack.push(stack.pop() * nextElem());
				else if (ch == '/')
					stack.push(stack.pop() / nextElem());
				else if (ch == '%')
					stack.push(stack.pop() % nextElem());
				else if (ch == '+' || ch == '-') {
					if (prevOp == '*' || prevOp == '/' || prevOp == '%' && stack.size() >= 2) {
						int b = stack.pop();
						int a = stack.pop();
						if (ch == '+')
							stack.push(a + b);
						else
							stack.push(a - b);
					} else if (ch == '+')
						stack.push(stack.pop() + nextElem());
					else
						stack.push(stack.pop() - nextElem());
				} else
					throw new Exception("unknown operator " + ch + " at " + (ind - 1));
				prevOp = ch;
			}
			if (prevOp == '+' || prevOp == '-' && stack.size() >= 2) {
				int b = stack.pop();
				int a = stack.pop();
				if (prevOp == '+')
					stack.push(a + b);
				else
					stack.push(a - b);
			}
			return stack.pop();
		}

		private int nextElem() throws Exception {
			char ch = str.charAt(ind);
			int neg = 1;
			if (ch == '-') {
				neg = -1;
				ch = str.charAt(++ind);
			}
			if (ch == '(') {
				int depth = 1;
				int pre = ++ind;
				while (depth > 0) {
					char chr = str.charAt(ind++);
					if (chr == '(')
						depth++;
					if (chr == ')')
						depth--;
				}
				return neg * new IntExp(pre, ind - 1).eval();
			}
			if (ch >= '0' && ch <= '9')
				return neg * readNumber();
			int pre = ind;
			while (ch != '+' && ch != '-' && ch != '*' && ch != '/' && ch != '%' && ind < p1)
				ch = str.charAt(++ind);
			return neg * (Integer) new RefObj(pre, ind).eval();
		}

		private int readNumber() {
			int ans = 0;
			while (ind < p1) {
				char chr = str.charAt(ind);
				if (chr < '0' || chr > '9')
					break;
				ind++;
				ans = ans * 10 + chr - '0';
			}
			return ans;
		}

	}

	private abstract static class RefElem extends Comp {

		public RefElem(int start, int end) {
			super(start, end);
		}

		public abstract Object eval(Object parent) throws Exception;
	}

	private class RefField extends RefElem {

		public RefField(int start, int end) {
			super(start, end);
		}

		@Override
		public Object eval(Object parent) throws Exception {
			if (str.charAt(p0) == '_' && parent != null)
				throw new Exception("global only allowed for bottom level");
			if (parent == null)
				if (str.charAt(p0) == '_')
					parent = ctx;
				else
					parent = obj;
			int ind = str.charAt(p0) == '_' ? p0 + 1 : p0;
			String name = str.substring(ind, p1);

			try {
				Field f = parent.getClass().getField(name);
				return f.get(parent);
			} catch (NoSuchFieldException nse) {
				return new IntExp(ind, p1).eval();
			}
		}
	}

	private class RefFunc extends RefElem {

		private final List<RefObj> list = new ArrayList<>();

		public RefFunc(int start, int end) {
			super(start, end);
		}

		@Override
		public Object eval(Object parent) throws Exception {
			if (str.charAt(p0) == '_' && parent != null)
				throw new Exception("global only allowed for bottom level: at " + p0);
			if (parent == null)
				if (str.charAt(p0) == '_')
					parent = ctx;
				else
					parent = obj;
			String name = str.substring(str.charAt(p0) == '_' ? p0 + 1 : p0, p1);
			Method[] ms = parent.getClass().getMethods();
			Object[] args = new Object[list.size()];
			for (int i = 0; i < args.length; i++)
				args[i] = list.get(i).eval();
			for (Method m : ms)
				if (m.getName().equals(name) && m.getParameterTypes().length == list.size())
					return m.invoke(parent, args);
			throw new Exception("function " + name + " not found for class " + parent.getClass());
		}
	}

	private class RefObj extends Elem {

		private final List<RefElem> list = new ArrayList<>();

		private RefObj(int start, int end) throws Exception {
			super(start, end);
			int pre = p0, i = p0;
			while (i < p1) {
				char ch = str.charAt(i++);
				if (ch == '.') {
					list.add(new RefField(pre, i - 1));
					pre = i;
				}
				if (ch == '(') {
					RefFunc func = new RefFunc(pre, i - 1);
					pre = i;
					int depth = 1;
					while (depth > 0) {
						if (i >= p1)
							throw new Exception("unfinished at " + i);
						char chr = str.charAt(i++);
						if (chr == '(')
							depth++;
						if (chr == ')') {
							depth--;
							if (depth == 0) {
								if (i - 1 > pre)
									func.list.add(new RefObj(pre, i - 1));
								pre = i;
							}
						}
						if (chr == ',' && depth == 1) {
							func.list.add(new RefObj(pre, i - 1));
							pre = i;
						}
					}
					list.add(func);
				}
			}
			if (pre < p1)
				list.add(new RefField(pre, p1));
		}

		@Override
		public void build(StringBuilder sb) throws Exception {
			sb.append(eval());
		}

		private Object eval() throws Exception {
			Object obj = null;
			for (RefElem e : list)
				obj = e.eval(obj);
			return obj;
		}

	}

	private class Root extends Cont {

		private Root(int start, int end) throws Exception {
			super(start, end);
			int pre = p0;
			int deepth = 0;
			for (int i = p0; i < p1; i++) {
				char ch = str.charAt(i);
				if (ch == '[') {
					if (deepth == 0 && i > pre) {
						list.add(new TextRef(pre, i));
						pre = i + 1;
					}
					deepth++;
				}
				if (ch == ']') {
					deepth--;
					if (deepth == 0 && i > pre) {
						list.add(new CodeBlock(pre, i));
						pre = i + 1;
					}
				}
			}
			if (pre < p1)
				list.add(new TextRef(pre, p1));
		}

	}

	private class TextPlain extends Elem {

		private TextPlain(int start, int end) {
			super(start, end);
		}

		@Override
		public void build(StringBuilder sb) {
			sb.append(str, p0, p1);
		}

	}

	private class TextRef extends Cont {

		private TextRef(int start, int end) throws Exception {
			super(start, end);
			int pre = p0;
			int depth = 0;
			for (int i = p0; i < p1; i++) {
				char ch = str.charAt(i);
				if (ch == '(') {
					if (depth == 0 && i > pre)
						list.add(new TextPlain(pre, i));
					if (depth == 0)
						pre = i + 1;
					depth++;
				}
				if (ch == ')') {
					depth--;
					if (depth == 0 && i > pre)
						list.add(new RefObj(pre, i));
					if (depth == 0)
						pre = i + 1;
				}
			}
			if (pre < p1)
				list.add(new TextPlain(pre, p1));
		}

	}

	@StaticPermitted
	private static final String[] MATCH = { ">=", "<=", "==", "!=", ">", "<" };

	public static String format(String pattern, Object obj, Object ctx) {
		StringBuilder sb = new StringBuilder();
		try {
			Formatter f = new Formatter(pattern, obj, ctx);
			f.root.build(sb);
		} catch (Exception e) {
			CommonStatic.ctx.noticeErr(e, ErrType.ERROR,
					"err during formatting " + pattern + " with " + JsonEncoder.encode(obj));
		}
		return sb.toString();
	}

	private final String str;
	private final Object obj;
	private final Object ctx;
	private final Root root;

	private Formatter(String pattern, Object object, Object context) throws Exception {
		str = pattern;
		obj = object;
		ctx = context;
		String err = check();
		if (err != null)
			throw new Exception(err);
		root = new Root(0, str.length());
	}

	private String check() {
		Stack<Integer> stack = new Stack<>();
		for (int i = 0; i < str.length(); i++) {
			char ch = str.charAt(i);
			if (ch == '(')
				stack.push(0);
			if (ch == '[')
				stack.push(1);
			if (ch == '{')
				stack.push(2);
			if (ch == ')' && (stack.isEmpty() || stack.pop() != 0))
				return "bracket ) unexpected at " + i;
			if (ch == ']' && (stack.isEmpty() || stack.pop() != 1))
				return "bracket ] unexpected at " + i;
			if (ch == '}' && (stack.isEmpty() || stack.pop() != 2))
				return "bracket } unexpected at " + i;
		}
		return stack.isEmpty() ? null : "unenclosed bracket: " + stack;
	}

}

File: util/lang/Formatter.kt
package common.util.lang

import common.CommonStatic
import common.io.assets.Admin.StaticPermitted
import common.io.json.JsonClass
import common.io.json.JsonEncoder
import common.io.json.JsonField
import common.pack.Context.ErrType
import common.pack.PackData
import common.util.lang.Formatter
import common.util.pack.Background
import java.util.*

class Formatter private constructor(private val str: String, private val obj: Any, private val ctx: Any) {
    @JsonClass
    class Context {
        val left = "("
        val right = ")"
        val sqleft = "["
        val sqright = "]"
        val crleft = "{"
        val crright = "}"

        @JsonField
        var isEnemy = false

        @JsonField
        var useSecond = false

        constructor() {}
        constructor(ene: Boolean, sec: Boolean) {
            isEnemy = ene
            useSecond = sec
        }

        fun abs(v: Int): String {
            return "" + Math.abs(v)
        }

        fun bg(id: PackData.Identifier<Background?>?): String {
            return PackData.Identifier.Companion.get<Background>(id).toString() + ""
        }

        fun dispTime(time: Int): String {
            return if (useSecond) toSecond(time) + "s" else time.toString() + "f"
        }

        fun entity(id: PackData.Identifier<*>?): String {
            return PackData.Identifier.Companion.get(id).toString() + ""
        }

        fun toSecond(time: Int): String {
            return "" + time * 100 / 30 / 100.0
        }
    }

    private inner class BoolElem(start: Int, end: Int) : Formatter.Comp(start, end) {
        @Throws(Exception::class)
        fun eval(): Boolean {
            for (i in Formatter.Companion.MATCH.indices) for (j in p0 until p1 - Formatter.Companion.MATCH.get(i).length + 1) if (test(i, j)) {
                val fi0 = IntExp(p0, j).eval()
                val fi1 = IntExp(j + Formatter.Companion.MATCH.get(i).length, p1).eval()
                if (i == 0) return fi0 >= fi1
                if (i == 1) return fi0 <= fi1
                if (i == 2) return fi0 == fi1
                if (i == 3) return fi0 != fi1
                if (i == 4) return fi0 > fi1
                if (i == 5) return fi0 < fi1
            }
            return RefObj(p0, p1).eval() as Boolean
        }

        private fun test(i: Int, j: Int): Boolean {
            for (k in 0 until Formatter.Companion.MATCH.get(i).length) if (str[j + k] != Formatter.Companion.MATCH.get(i).get(k)) return false
            return true
        }
    }

    private inner class BoolExp(start: Int, end: Int) : Formatter.Comp(start, end) {
        private var ind: Int

        @Throws(Exception::class)
        fun eval(): Boolean {
            val stack = Stack<Boolean>()
            stack.push(nextElem())
            while (ind < p1) {
                val ch = str[ind++]
                if (ch == '&') stack.push(stack.pop() and nextElem()) else if (ch == '|') stack.push(nextElem()) else throw Exception("unknown operator " + ch + " at " + (ind - 1))
            }
            var b = false
            for (bi in stack) b = b or bi
            return b
        }

        @Throws(Exception::class)
        private fun nextElem(): Boolean {
            var ch = str[ind]
            val neg = ch == '!'
            if (neg) ch = str[++ind]
            if (ch == '!') throw Exception("double ! at $ind")
            if (ch == '(') {
                var depth = 1
                val pre = ++ind
                while (depth > 0) {
                    val chr = str[ind++]
                    if (chr == '(') depth++
                    if (chr == ')') depth--
                }
                return neg xor BoolExp(pre, ind - 1).eval()
            }
            val pre = ind
            while (ch != '&' && ch != '|' && ind < p1) ch = str[++ind]
            return neg xor BoolElem(pre, ind).eval()
        }

        init {
            ind = p0
        }
    }

    private inner class Code private constructor(private val cond: BoolExp, private val data: Formatter.Root) : IElem {
        @Throws(Exception::class)
        override fun build(sb: StringBuilder) {
            if (cond.eval()) data.build(sb)
        }
    }

    private inner class CodeBlock(start: Int, end: Int) : Cont(start, end) {
        init {
            var i = p0
            while (i < p1) {
                val ch = str[i++]
                if (ch == '(') {
                    var depth = 1
                    var pre = i
                    while (depth > 0) {
                        if (i >= p1) throw Exception("unfinished at $i")
                        val chr = str[i++]
                        if (chr == '(') depth++
                        if (chr == ')') depth--
                    }
                    val cond = BoolExp(pre, i - 1)
                    while (str[i++] != '{') if (i >= p1) throw Exception("unfinished at $i")
                    depth = 1
                    pre = i
                    while (depth > 0) {
                        if (i >= p1) throw Exception("unfinished at $i")
                        val chr = str[i++]
                        if (chr == '{') depth++
                        if (chr == '}') depth--
                    }
                    val data: Formatter.Root = Formatter.Root(pre, i - 1)
                    list.add(Formatter.Code(cond, data))
                }
            }
        }
    }

    private abstract inner class Comp(val p0: Int, val p1: Int)
    private abstract inner class Cont(start: Int, end: Int) : Elem(start, end) {
        val list: MutableList<IElem> = ArrayList()
        @Throws(Exception::class)
        override fun build(sb: StringBuilder) {
            for (e in list) e.build(sb)
        }
    }

    private abstract inner class Elem(start: Int, end: Int) : Formatter.Comp(start, end), IElem
    private interface IElem {
        @Throws(Exception::class)
        fun build(sb: StringBuilder)
    }

    private inner class IntExp(start: Int, end: Int) : Formatter.Comp(start, end) {
        private var ind: Int

        @Throws(Exception::class)
        fun eval(): Int {
            val stack = Stack<Int>()
            var opera = '\u0000'
            stack.push(nextElem())
            while (ind < p1) {
                val ch = str[ind++]
                if (ch == '*') stack.push(stack.pop() * nextElem()) else if (ch == '/') stack.push(stack.pop() / nextElem()) else if (ch == '%') stack.push(stack.pop() % nextElem()) else if (ch == '+' || ch == '-') {
                    if (opera != ' ') {
                        val b = stack.pop()
                        val a = stack.pop()
                        if (opera == '+') stack.push(a + b) else stack.push(a - b)
                    }
                    stack.push(nextElem())
                    opera = ch
                } else throw Exception("unknown operator " + ch + " at " + (ind - 1))
            }
            if (opera != '\u0000') {
                val b = stack.pop()
                val a = stack.pop()
                if (opera == '+') stack.push(a + b) else stack.push(a - b)
            }
            return stack.pop()
        }

        @Throws(Exception::class)
        private fun nextElem(): Int {
            var ch = str[ind]
            var neg = 1
            if (ch == '-') {
                neg = -1
                ch = str[++ind]
            }
            if (ch == '(') {
                var depth = 1
                val pre = ++ind
                while (depth > 0) {
                    val chr = str[ind++]
                    if (chr == '(') depth++
                    if (chr == ')') depth--
                }
                return neg * IntExp(pre, ind - 1).eval()
            }
            if (ch >= '0' && ch <= '9') return neg * readNumber()
            val pre = ind
            while (ch != '+' && ch != '-' && ch != '*' && ch != '/' && ch != '%' && ind < p1) ch = str[++ind]
            return neg * (RefObj(pre, ind).eval() as Int?)!!
        }

        @Throws(Exception::class)
        private fun readNumber(): Int {
            var ans = 0
            while (ind < p1) {
                val chr = str[ind]
                if (chr < '0' || chr > '9') break
                ind++
                ans = ans * 10 + chr.toInt() - '0'.toInt()
            }
            return ans
        }

        init {
            ind = p0
        }
    }

    private abstract inner class RefElem(start: Int, end: Int) : Formatter.Comp(start, end) {
        @Throws(Exception::class)
        abstract fun eval(parent: Any?): Any?
    }

    private inner class RefField(start: Int, end: Int) : RefElem(start, end) {
        @Throws(Exception::class)
        override fun eval(parent: Any?): Any? {
            var parent = parent
            if (str[p0] == '_' && parent != null) throw Exception("global only allowed for bottom level")
            if (parent == null) parent = if (str[p0] == '_') ctx else obj
            val name = str.substring(if (str[p0] == '_') p0 + 1 else p0, p1)
            val f = parent.javaClass.getField(name)
            return f[parent]
        }
    }

    private inner class RefFunc(start: Int, end: Int) : RefElem(start, end) {
        val list: MutableList<RefObj> = ArrayList()
        @Throws(Exception::class)
        override fun eval(parent: Any?): Any? {
            var parent = parent
            if (str[p0] == '_' && parent != null) throw Exception("global only allowed for bottom level: at $p0")
            if (parent == null) parent = if (str[p0] == '_') ctx else obj
            val name = str.substring(if (str[p0] == '_') p0 + 1 else p0, p1)
            val ms = parent.javaClass.methods
            val args = arrayOfNulls<Any>(list.size)
            for (i in args.indices) args[i] = list[i].eval()
            for (m in ms) if (m.name == name && m.parameterCount == list.size) return m.invoke(parent, *args)
            throw Exception("function " + name + " not found for class " + parent.javaClass)
        }
    }

    private inner class RefObj(start: Int, end: Int) : Elem(start, end) {
        private val list: MutableList<RefElem> = ArrayList()
        @Throws(Exception::class)
        override fun build(sb: StringBuilder) {
            sb.append("" + eval())
        }

        @Throws(Exception::class)
        fun eval(): Any? {
            var obj: Any? = null
            for (e in list) obj = e.eval(obj)
            return obj
        }

        init {
            var pre = p0
            var i = p0
            while (i < p1) {
                val ch = str[i++]
                if (ch == '.') {
                    list.add(RefField(pre, i - 1))
                    pre = i
                }
                if (ch == '(') {
                    val func = RefFunc(pre, i - 1)
                    pre = i
                    var depth = 1
                    while (depth > 0) {
                        if (i >= p1) throw Exception("unfinished at $i")
                        val chr = str[i++]
                        if (chr == '(') depth++
                        if (chr == ')') {
                            depth--
                            if (depth == 0) {
                                if (i - 1 > pre) func.list.add(RefObj(pre, i - 1))
                                pre = i
                            }
                        }
                        if (chr == ',' && depth == 1) {
                            func.list.add(RefObj(pre, i - 1))
                            pre = i
                        }
                    }
                    list.add(func)
                }
            }
            if (pre < p1) list.add(RefField(pre, p1))
        }
    }

    private inner class Root private constructor(start: Int, end: Int) : Cont(start, end) {
        init {
            var pre = p0
            var deepth = 0
            for (i in p0 until p1) {
                val ch = str[i]
                if (ch == '[') {
                    if (deepth == 0 && i > pre) {
                        list.add(TextRef(pre, i))
                        pre = i + 1
                    }
                    deepth++
                }
                if (ch == ']') {
                    deepth--
                    if (deepth == 0 && i > pre) {
                        list.add(CodeBlock(pre, i))
                        pre = i + 1
                    }
                }
            }
            if (pre < p1) list.add(TextRef(pre, p1))
        }
    }

    private inner class TextPlain(start: Int, end: Int) : Elem(start, end) {
        override fun build(sb: StringBuilder) {
            sb.append(str, p0, p1)
        }
    }

    private inner class TextRef(start: Int, end: Int) : Cont(start, end) {
        init {
            var pre = p0
            var depth = 0
            for (i in p0 until p1) {
                val ch = str[i]
                if (ch == '(') {
                    if (depth == 0 && i > pre) list.add(TextPlain(pre, i))
                    if (depth == 0) pre = i + 1
                    depth++
                }
                if (ch == ')') {
                    depth--
                    if (depth == 0 && i > pre) list.add(RefObj(pre, i))
                    if (depth == 0) pre = i + 1
                }
            }
            if (pre < p1) list.add(TextPlain(pre, p1))
        }
    }

    private val root: Formatter.Root
    private fun check(): String? {
        val stack = Stack<Int>()
        for (i in 0 until str.length) {
            val ch = str[i]
            if (ch == '(') stack.push(0)
            if (ch == '[') stack.push(1)
            if (ch == '{') stack.push(2)
            if (ch == ')' && (stack.isEmpty() || stack.pop() != 0)) return "bracket ) unexpected at $i"
            if (ch == ']' && (stack.isEmpty() || stack.pop() != 1)) return "bracket ] unexpected at $i"
            if (ch == '}' && (stack.isEmpty() || stack.pop() != 2)) return "bracket } unexpected at $i"
        }
        return if (stack.isEmpty()) null else "unenclosed bracket: $stack"
    }

    companion object {
        @StaticPermitted
        private val MATCH = arrayOf(">=", "<=", "==", "!=", ">", "<")
        fun format(pattern: String, obj: Any?, ctx: Any?): String {
            val sb = StringBuilder()
            try {
                val f = common.util.lang.Formatter(pattern, obj, ctx)
                f.root.build(sb)
            } catch (e: Exception) {
                CommonStatic.ctx.noticeErr(e, ErrType.ERROR,
                        "err during formatting " + pattern + " with " + JsonEncoder.Companion.encode(obj))
            }
            return sb.toString()
        }
    }

    init {
        val err = check()
        if (err != null) throw Exception(err)
        root = Formatter.Root(0, str.length)
    }
}
